---
title: "DESeq2_tutorial"
author: "Guannan Shen"
date: "August 15, 2018"
output:
  html_document:
    number_sections: yes
    theme: united
    toc: yes
    toc_depth: 5
    toc_float: yes
  pdf_document:
    number_sections: yes
    toc: yes
    toc_depth: 5
  word_document:
    toc: yes
    toc_depth: '5'
---
# Analyzing RNA-seq data with DESeq2

```{r setup, echo=TRUE, message=FALSE, warning=FALSE, prompt=TRUE, tidy=TRUE, include=FALSE}
######################
## Set up workspace
######################
rm(list = ls())
options(stringsAsFactors = F)
library(rmarkdown)
library(knitr)
opts_chunk$set(tidy.opts=list(width.cutoff=60),tidy=TRUE)
knitr::opts_chunk$set(engine = "R")
getwd()
```

## Abstract
An important analysis question, for the detection of differentially expressed genes, is the quantification and statistical inference of systematic changes between conditions, as compared to within-condition variability.  
  
The package DESeq2 provides methods to test for differential expression by use of **negative binomial generalized linear models**; the estimates of dispersion and logarithmic fold changes incorporate data-driven prior distributions. We will start from the FASTQ files, show how these were aligned to the reference genome, and prepare a count matrix which tallies the number of RNA-seq reads/fragments within each gene for each sample. We will perform exploratory data analysis (EDA) for quality assessment and to explore the relationship between samples, perform differential gene expression analysis, and visually explore the results.  

## Introduction 
Through scheduled releases every 6 months, the Bioconductor project ensures that all the packages within a release will work together in harmony (hence the "conductor" metaphor).  
The data used in this workflow is stored in the **airway package** that summarizes an RNA-seq experiment wherein airway smooth muscle cells were treated with dexamethasone, a synthetic glucocorticoid steroid with anti-inflammatory effects (Himes et al. 2014). In the experiment, four primary human airway smooth muscle cell lines were treated with 1 micromolar dexamethasone for 18 hours. For each of the **four cell lines, we have a treated and an untreated sample**. 

## Preparing Count matrices
As input, the count-based statistical methods, such as **DESeq2, edgeR, limma with the voom method, DSS, EBSeq, baySeq**, expect input data as obtained in the form of a matrix of integer values. The value in the i-th row and the j-th column of the matrix tells how many reads (or fragments, for paired-end RNA-seq) have been assigned to gene i in sample j. The number of reads mapped to a given gene or transcript is considered to be the estimate of the expression level of that feature using this technology.  
One of the issues is that the expression quantification from short reads using RNA-seq data depends on the length of the features; longer features usually produce more reads. Sequence depth can be calculated as $N × L/G$, where N is the number of reads, L is the average read length and G is the length of the original genome. This also is equivalent to the percentage of genome covered by reads and the average number of times a base is read. Higher coverage can improve the power to identify differential expression using RNA-seq data. Overall read count for a samle is the library size. 
It is important to never provide counts that were pre-normalized for sequencing depth/library size, as the statistical model is most powerful when applied to un-normalized counts, and is designed to account for library size differences internally.

### For alignment and count RNA-seq fragments 

#### Align, or not: transcript abundances
Newer and faster alternative pipeline is to use **transcript abundance quantification** methods, such as **Salmon, Sailfish, kallisto, RSEM**, to estimate abundances without aligning reads, followed by the tximport package for assembling estimated count and offset matrices for use with Bioconductor differential gene expression packages. This approach corrects for any potential changes in gene length across samples (e.g. from differential isoform usage).  
The computational analysis of an RNA-seq experiment begins from the FASTQ files that contain the nucleotide sequence of each read and a quality score at each position. These reads must first be aligned to a reference genome or transcriptome, or the abundances and estimated counts per transcript can be estimated without alignment, as described above. In either case, it is important to know if the sequencing experiment was **single-end or paired-end**. 


```{r}
```

## References
1. Analyzing RNA-seq data with DESeq2, Michael I. Love, Simon Anders, and Wolfgang Huber, http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html  
2. RNA-seq workflow: gene-level exploratory analysis and differential expression, Michael I. Love, Simon Anders, Vladislav Ki3 and Wolfgang Huber, http://master.bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html 
3. Statistical Analysis of Next Generation Sequencing Data

