---
title: "Single Cell RNA-Seq"
author: "Guannan Shen"
date: "December 11, 2018"
output: 
  pdf_document:
    latex_engine: lualatex
    number_sections: yes
    toc: yes
    toc_depth: 5
  word_document:
    toc: yes
    toc_depth: '5'
  html_document:
    number_sections: yes
    theme: united
    toc: yes
    toc_depth: 5
    toc_float: yes
---

```{r setup, include=FALSE, cache = FALSE}
require("knitr")
opts_chunk$set(tidy.opts=list(width.cutoff=60),tidy=TRUE)
opts_chunk$set(engine = "R")
knitr::opts_chunk$set(echo = T)
knitr::opts_chunk$set(message = F)
knitr::opts_chunk$set(warning = F)
## setting wd in DELL
## opts_knit$set(root.dir = "~/Stats/CIDA_OMICs/CIDA_OMICS/7659Stats_Genetics/HW5/")
## setting working directory in asus 
## opts_knit$set(root.dir = "C:/Users/hithr/Documents/Stats/CIDA_OMICs/7659Stats_Genetics/HW9/") 
## setting working directory in ubuntu
opts_knit$set(root.dir = "~/Documents/Stats/CIDA_OMICs/7659Stats_Genetics/Proj_Single_Cell/DataRaw/")
                                                 
## cache = F, if cache = T, will not revaluate code chunk everytime
## double or more space to insert a line break
```


```{r libs}
## set up workspace
library(Seurat)
library(knitr)
library(tidyverse)
library(magrittr)
library(stats)
library(data.table)
options(stringsAsFactors = F)
options(dplyr.width = Inf)
getwd()
## not in function
'%nin%' <- Negate('%in%')

# ######## clean memory ######################
# rm(list = ls())
# gc()
# slotNames(x)
# getSlots(x)

```



# Setup the Seurat Object


```{r setup}
# Sample GSM3374613: **Original single cell RNA-seq library** from the 10x genomics 3' end library prep. PBMCs were isolated from a health human donor. 
ctrl.data <- read.table("~/Documents/Stats/7659Genetics/Proj/GSM3374611_original_10x_umis.tsv.gz", sep = "\t", header = T)
dim(ctrl.data)
head(ctrl.data[, 1:6])
## check the sparsity
genelevelSparse_ctrl <-  apply(ctrl.data, 1, function(x){
  sum(x == 0)/length(x)
})
sum(genelevelSparse_ctrl != 0)
## boxplot(log2(ctrl.data), main = "The Counts of Single Cell RNA-Seq", ylab= "log2(Counts)")
# Sample GSM3374614 PBMCs were isolated from a health human donor and a LNA probe was used to hybridize selected cells for targeted resequencing.
stim.data <- as.data.frame(as.matrix(fread("~/Documents/Stats/CIDA_OMICs/7659Stats_Genetics/Proj_Single_Cell/DataRaw/GSM3374614_mkcell_pulldown_umis.tsv")))
dim(stim.data)
genelevelSparse_stim <-apply(stim.data, 1, function(x){
  sum(x == 0)/length(x)
}) 
sum(genelevelSparse_stim != 0)
# Initialize the Seurat object with the raw (non-normalized data).  Keep all
# genes expressed in >= 3 cells (~0.1% of the data). Keep all cells with at
# least 200 detected genes
pbmc <- CreateSeuratObject(raw.data = ctrl.data, min.cells = 3, min.genes = 200, 
    project = "10X_PBMC")

```







